drivers: usb: udc: numaker: support m333x hsusbd

ccli8 opened the pull request:
This adds support for Nuvoton NuMaker m333x HSUSBD and makes other fixes as below:

1. Add config for disabling USB on unplug for soc supporting this feature
2. Fix control out failure with FIFO flush timing in HSUSBD
3. Fix compile error on m2l31x
4. Fix submit condition to meet updated msc class driver which requests submit data to be multiple of sector size

ccli8 committed:
drivers: usb: udc: numaker: support m333x hsusbd

Add support for Nuvoton NuMaker M3331 series HSUSBD device driver

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: add config for disabling USB on unplug

Some SoC series allow USB controller to be disabled temporarily
on unplug, and then enabled back on re-plug. Add support for such
configuration option in DT and UDC driver.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: fix compile error on m2l31x

With USBD and HSUSBD re-organized, SoC must have USBD and HSUSBD at
the same time to pass compile. However, M2L31 just has USBD and meets
compile error with HSUSBD unsupported. To avoid adding too many
conditional (#if condition) code for the error, create one dummy
HSUSBD device definition for M2L31 to pass compile.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: fix retrieval of setup packet

Change to retrieve Setup packet by the copy which is readily copied
from H/W register in ISR

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: fix control out failure in hsusbd

For HSUSBD, Control transfer with Data OUT stage will break. This is
caused by FIFO flush which cannot be done timely even though in Setup
token interrupt handler. For this, error recovery with FIFO is not
done in Setup token/packet handler and rely on USB reset handler to do
it as catch-all.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

ccli8 committed:
drivers: usb: udc: numaker: fix submit condition

Fix submit condition for non-control transfer. Do submit when any
of the following conditions is met:
1. Transfer buffer (net_buf) is full
2. Last packet size is less than mps
3. Isochronous transfer

USB mass may request C1 for sector size (512 bytes)-aligned and C2
for CBW (13 bytes).

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>

josuah commented on dts/bindings/usb/nuvoton,numaker-usbd.yaml:
```diff
@@ -31,3 +31,9 @@ properties:
+    type: boolean
+    description: |
+      Some soc series allow USB function to be disabled temporarily on USB unplug. It can enable
+      back on USB re-plug.
```
Same feedback as https://github.com/zephyrproject-rtos/zephyr/pull/101460/changes#r2669016088

ccli8 replied:
Same as above

josuah reviewed (dismissed):
I do not know this hardware, but best I can say is I did not find issues regarding the USB device driver API.

tmon-nordic commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -2100,7 +2094,11 @@ static int numaker_usbd_msg_handle_setup(const struct device *dev, struct numake
+		 * is done in Setup token ISR (HSUSBD_CEPINTSTS_SETUPTKIF_Msk),
+		 * it can still be not timely. For this, error recovery with FIFO
+		 * is not done in-place here and rely on USB reset handler to do it
+		 * as catch-all.
```
This sounds like some serious hardware limitation. Is there really not a register access sequence that would guarantee proper operation for a Control Write (with Data OUT stage) transfer regardless of "software timing"? It sounds strange that there would be inevitable race condition, where it would have to rely on USB reset handler (I assume host issues reset after multiple control transfer timeouts) to recover.

ccli8 replied:
About this and above concerns on consecutive Control transfers, due to the hardware limitation, there is no proper register control to address them, and keep  current implementation.

tmon-nordic commented on dts/bindings/usb/nuvoton,numaker-hsusbd.yaml:
```diff
@@ -25,3 +25,9 @@ properties:
+    type: boolean
+    description: |
+      Some soc series allow USB functionality to be disabled temporarily on USB unplug.
+      It can enable back on USB re-plug.
```
I think this is not quite the format expected in devicetree binding descriptions, but I am not sure what the proper text would be. I think something more like "USB controller can be disabled when USB cable is unplugged." would fit better, but don't take my word on it.

ccli8 commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -2456,7 +2458,18 @@ static int numaker_usbd_msg_handle_out(const struct device *dev, struct numaker_
-	} else {
+	} else if ((net_buf_tailroom(buf) == 0) || (data_len < ep_cfg->mps) ||
+		   (ep_type == USB_EP_TYPE_ISO)) {
+		/* Fix submit condition for non-control transfer
```
In previous implementation, BULK OUT data is submitted to stack immediately on receive. For MSC class driver, the data size can be 31 bytes (CBW) or 64 bytes (sector data), assuming wMaxPacketSize is 64 bytes. This will fail with updated MSC , in which sector data must collect to multiple of 512 bytes (net_buf being full) before submit. So there are the submit conditions C1 and C2 for MSC.

I am not sure if stack has definition on size conditions to submit DATA OUT data. It is appreciated if someone could hint.

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -71,6 +71,143 @@ LOG_MODULE_REGISTER(udc_numaker, CONFIG_UDC_DRIVER_LOG_LEVEL);
+#if defined(CONFIG_SOC_SERIES_M2L31X)
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(nuvoton_numaker_hsusbd) == 0,
+	     "The SoC series should have no HSUSBD");
```
What is the point of this BUILD_ASSERT?

jfischer-no commented on drivers/usb/udc/udc_numaker.c:
```diff
@@ -71,6 +71,143 @@ LOG_MODULE_REGISTER(udc_numaker, CONFIG_UDC_DRIVER_LOG_LEVEL);
 #endif
 #endif
 
+/* Dummy HSUSBD device definition
```
Well, that is why I asked. Or perhaps you can add a common header file for this controller type to avoid 140 lines of dummy defines?

jfischer-no replied:
You could add drivers/usb/udc/udc_numaker.h and move 
`#include <soc.h>`, `#include <NuMicro.h>` and these dummy defines there. @fabiobaltieri, what are your thoughts on this?

ccli8 committed:
usb: udc: numaker: move dummy usbd/hsusbd to separate header

Move dummy USBD/HSUSBD register defines to separate header file
so that the code can be more clear.

Signed-off-by: Chun-Chieh Li <ccli8@nuvoton.com>
